src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -0,0 +1,281 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Manages native memory allocations made by JNI.
+ */
+public class NativeMemoryCacheManager implements Closeable {
+
+    public static String GRAPH_COUNT = "graph_count";
+
+    private static Logger logger = LogManager.getLogger(NativeMemoryCacheManager.class);
+    private static NativeMemoryCacheManager INSTANCE;
+
+    private Cache<String, NativeMemoryAllocation> cache;
+    private ExecutorService executor;
+    private AtomicBoolean cacheCapacityReached;
+    private long maxWeight;
+
+    NativeMemoryCacheManager() {
+        this.executor = Executors.newSingleThreadExecutor();
+        this.cacheCapacityReached = new AtomicBoolean(false);
+        this.maxWeight = Long.MAX_VALUE;
+        initialize();
+    }
+
+    /**
+     * Make sure we just have one instance of cache.
+     *
+     * @return NativeMemoryCacheManager instance
+     */
+    public static synchronized NativeMemoryCacheManager getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new NativeMemoryCacheManager();
+        }
+        return INSTANCE;
+    }
+
+    private void initialize() {
+        CacheBuilder<String, NativeMemoryAllocation> cacheBuilder = CacheBuilder.newBuilder()
+                .recordStats()
+                .concurrencyLevel(1)
+                .removalListener(this::onRemoval);
+
+        if(KNNSettings.state().getSettingValue(KNNSettings.KNN_MEMORY_CIRCUIT_BREAKER_ENABLED)) {
+            maxWeight = KNNSettings.getCircuitBreakerLimit().getKb();
+            cacheBuilder.maximumWeight(maxWeight).weigher((k, v) -> (int)v.getSizeInKb());
+        }
+
+        if(KNNSettings.state().getSettingValue(KNNSettings.KNN_CACHE_ITEM_EXPIRY_ENABLED)) {
+            long expiryTime = ((TimeValue) KNNSettings.state()
+                    .getSettingValue(KNNSettings.KNN_CACHE_ITEM_EXPIRY_TIME_MINUTES)).getMinutes();
+            cacheBuilder.expireAfterAccess(expiryTime, TimeUnit.MINUTES);
+        }
+
+        cacheCapacityReached = new AtomicBoolean(false);
+
+        cache = cacheBuilder.build();
+    }
+
+    /**
+     * Evicts all entries from the cache and rebuilds.
+     */
+    public synchronized void rebuildCache() {
+        logger.info("KNN Cache rebuilding.");
+        executor.execute(() -> {
+            cache.invalidateAll();
+            initialize(); }
+        );
+    }
+
+    @Override
+    public void close() {
+        executor.shutdown();
+    }
+
+    /**
+     * Getter for current cache weight in Kilobytes.
+     *
+     * @return current weight of the cache
+     */
+    public long getCacheWeightInKilobytes() {
+        return cache.asMap().values().stream().mapToLong(NativeMemoryAllocation::getSizeInKb).sum();
```

Comment:
It is valid in the sense that it will not break. It will tell you how much native memory the cache is consuming (regardless of whether there is a weight limit or not). However, I could see where the naming is misleading. Do you have a suggestion for another?

Reply:
How about the following?

getCacheWeightInKilobytes => getCacheSizeInKilobytes

getWeightForIndexInKilobytes => getIndexSizeInKilobytes

Reply:
Right, so I chose "weight" instead of "size" for these functions because there is a distinction made in Guava cache where size is the number of elements in the cache and weight is sum of some weigher function applied on each element in the cache.

That being said, to the user, I dont think it matters. I can switch to size.